File: scripts/crop_logos.py
import os
from PIL import Image, ImageChops

clients_dir = os.path.join(os.path.dirname(__file__), '..', 'assets', 'images', 'clients')

def trim(im, border_color=None, tolerance=25):
    if im.mode != 'RGBA':
        im = im.convert('RGBA')
    
    # Get corner pixel color if not provided
    if border_color is None:
        border_color = im.getpixel((0, 0))
    
    bg = Image.new(im.mode, im.size, border_color)
    diff = ImageChops.difference(im, bg)
    
    # Also handle transparent pixels
    # Find bounding box
    bbox = None
    # Check non-background pixels
    width, height = im.size
    min_x, min_y, max_x, max_y = width, height, 0, 0
    has_content = False
    
    pixels = im.load()
    br, bg_g, bb, ba = border_color
    
    for y in range(height):
        for x in range(width):
            r, g, b, a = pixels[x, y]
            is_bg = False
            if a < 30:
                is_bg = True
            elif ba > 200 and abs(r - br) < tolerance and abs(g - bg_g) < tolerance and abs(b - bb) < tolerance:
                is_bg = True
            elif r > 240 and g > 240 and b > 240 and abs(r - g) < 10 and abs(r - b) < 10:
                # White/near-white
                is_bg = True
            
            if not is_bg:
                has_content = True
                if x < min_x: min_x = x
                if x > max_x: max_x = x
                if y < min_y: min_y = y
                if y > max_y: max_y = y
                
    if has_content and max_x >= min_x and max_y >= min_y:
        pad = 8
        min_x = max(0, min_x - pad)
        min_y = max(0, min_y - pad)
        max_x = min(width - 1, max_x + pad)
        max_y = min(height - 1, max_y + pad)
        return (min_x, min_y, max_x + 1, max_y + 1)
    return None

def process_all():
    files = os.listdir(clients_dir)
    for fname in files:
        fpath = os.path.join(clients_dir, fname)
        if not os.path.isfile(fpath):
            continue
        try:
            im = Image.open(fpath)
            orig_size = im.size
            
            # Special handling for maxira_solar (cream background)
            if 'maxira' in fname.lower():
                # Corner color is cream
                bbox = trim(im, border_color=im.getpixel((0, 0)), tolerance=20)
            elif 'sunmoon' in fname.lower():
                # Dark background
                # Let's find content on black
                width, height = im.size
                pixels = im.convert('RGBA').load()
                min_x, min_y, max_x, max_y = width, height, 0, 0
                has_content = False
                for y in range(height):
                    for x in range(width):
                        r, g, b, a = pixels[x, y]
                        # Look for bright pixels (orange sun / white text)
                        if r > 50 or g > 50 or b > 50:
                            has_content = True
                            if x < min_x: min_x = x
                            if x > max_x: max_x = x
                            if y < min_y: min_y = y
                            if y > max_y: max_y = y
                if has_content:
                    pad = 12
                    min_x = max(0, min_x - pad)
                    min_y = max(0, min_y - pad)
                    max_x = min(width - 1, max_x + pad)
                    max_y = min(height - 1, max_y + pad)
                    bbox = (min_x, min_y, max_x + 1, max_y + 1)
                else:
                    bbox = None
            else:
                bbox = trim(im)
            
            if bbox:
                cropped = im.crop(bbox)
                print(f"Cropped {fname}: {orig_size} -> {cropped.size} (bbox: {bbox})")
                
                # If PNG, preserve transparency
                if fname.lower().endswith('.png'):
                    cropped.save(fpath, 'PNG', optimize=True)
                else:
                    # Save as RGB JPEG
                    if cropped.mode != 'RGB':
                        # Create white background
                        bg = Image.new('RGB', cropped.size, (255, 255, 255))
                        if cropped.mode == 'RGBA':
                            bg.paste(cropped, mask=cropped.split()[3])
                        else:
                            bg.paste(cropped)
                        bg.save(fpath, 'JPEG', quality=95)
                    else:
                        cropped.save(fpath, 'JPEG', quality=95)
            else:
                print(f"No crop for {fname}")
        except Exception as e:
            print(f"Error processing {fname}: {e}")

if __name__ == '__main__':
    process_all()

File: scripts/crop_logos.php
<?php
/**
 * Auto-crop whitespace from all logo images in assets/images/clients/
 */
$dir = __DIR__ . '/../assets/images/clients';
$files = scandir($dir);

foreach ($files as $file) {
    if ($file === '.' || $file === '..') continue;
    $filePath = "$dir/$file";
    $ext = strtolower(pathinfo($filePath, PATHINFO_EXTENSION));

    if (!in_array($ext, ['png', 'jpg', 'jpeg'])) continue;

    echo "Processing $file...\n";

    if ($ext === 'png') {
        $img = imagecreatefrompng($filePath);
    } else {
        $img = imagecreatefromjpeg($filePath);
    }

    if (!$img) {
        echo "Failed to load $file\n";
        continue;
    }

    $width = imagesx($img);
    $height = imagesy($img);

    // Sample corner pixels to determine background color
    $cornerColors = [];
    $corners = [[0, 0], [$width - 1, 0], [0, $height - 1], [$width - 1, $height - 1]];
    foreach ($corners as $c) {
        $rgb = imagecolorat($img, $c[0], $c[1]);
        $colors = imagecolorsforindex($img, $rgb);
        $cornerColors[] = $colors;
    }

    // Determine if background is transparent or near-white / solid
    $bgR = $cornerColors[0]['red'];
    $bgG = $cornerColors[0]['green'];
    $bgB = $cornerColors[0]['blue'];
    $bgA = $cornerColors[0]['alpha'];

    $minX = $width;
    $minY = $height;
    $maxX = 0;
    $maxY = 0;

    for ($y = 0; $y < $height; $y++) {
        for ($x = 0; $x < $width; $x++) {
            $rgb = imagecolorat($img, $x, $y);
            $c = imagecolorsforindex($img, $rgb);

            $isBg = false;
            if ($bgA > 100 && $c['alpha'] > 100) {
                $isBg = true;
            } elseif ($c['alpha'] > 120) {
                $isBg = true; // transparent
            } else {
                // If background is white/near-white or near the corner color
                $diff = abs($c['red'] - $bgR) + abs($c['green'] - $bgG) + abs($c['blue'] - $bgB);
                if ($diff < 25 && ($bgR > 220 || $bgG > 220 || $bgB > 220)) {
                    $isBg = true;
                } elseif ($c['red'] > 245 && $c['green'] > 245 && $c['blue'] > 245) {
                    $isBg = true; // almost pure white
                }
            }

            if (!$isBg) {
                if ($x < $minX) $minX = $x;
                if ($x > $maxX) $maxX = $x;
                if ($y < $minY) $minY = $y;
                if ($y > $maxY) $maxY = $y;
            }
        }
    }

    if ($maxX >= $minX && $maxY >= $minY) {
        $pad = 10;
        $minX = max(0, $minX - $pad);
        $minY = max(0, $minY - $pad);
        $maxX = min($width - 1, $maxX + $pad);
        $maxY = min($height - 1, $maxY + $pad);

        $cropW = $maxX - $minX + 1;
        $cropH = $maxY - $minY + 1;

        echo "  Cropping $file: original {$width}x{$height} -> cropped {$cropW}x{$cropH} (bounds: $minX, $minY to $maxX, $maxY)\n";

        $cropped = imagecreatetruecolor($cropW, $cropH);
        
        if ($ext === 'png') {
            imagealphablending($cropped, false);
            imagesavealpha($cropped, true);
            $transparent = imagecolorallocatealpha($cropped, 255, 255, 255, 127);
            imagefilledrectangle($cropped, 0, 0, $cropW, $cropH, $transparent);
        } else {
            $white = imagecolorallocate($cropped, 255, 255, 255);
            imagefilledrectangle($cropped, 0, 0, $cropW, $cropH, $white);
        }

        imagecopy($cropped, $img, 0, 0, $minX, $minY, $cropW, $cropH);

        if ($ext === 'png') {
            imagepng($cropped, $filePath, 9);
        } else {
            imagejpeg($cropped, $filePath, 95);
        }
        imagedestroy($cropped);
    } else {
        echo "  No crop needed or bounds not found for $file\n";
    }

    imagedestroy($img);
}

echo "All logos cropped successfully!\n";
